plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataCache.java

```diff
@@ -0,0 +1,586 @@
+    {
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) tableHandle;
+        SchemaTableWrappedCacheKey<ConnectorTableHandle> key = new SchemaTableWrappedCacheKey<>(icebergTableHandle.getSchemaTableName(), tableHandle);
+        return get(tableStatsCache, key);
+    }
+
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) tableHandle;
+        SchemaTableWrappedCacheKey<ConnectorTableHandle> key = new SchemaTableWrappedCacheKey<>(icebergTableHandle.getSchemaTableName(), tableHandle);
+        return get(tableMetadataCache, key);
+    }
+
+    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
+    {
+        return get(tableListingCache, schemaName);
+    }
+
+    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        Table icebergTable = get(tableCache, table.getSchemaTableName());
+        return getColumns(icebergTable.schema(), typeManager).stream()
+                .collect(toImmutableMap(IcebergColumnHandle::getName, identity()));
+    }
+
+    public Map<SchemaTableName, ConnectorViewDefinition> getViews(ConnectorSession session, Optional<String> schemaName)
+    {
+        return get(viewListingCache, schemaName);
+    }
+
+    public Optional<ConnectorViewDefinition> getView(ConnectorSession session, SchemaTableName viewName)
+    {
+        return get(viewCache, viewName);
+    }
+
+    private List<String> listCachedSchemaNames(String key)
+    {
+        IcebergMetadata icebergMetadata = metadataFactory.create(null);
+        return icebergMetadata.listSchemaNames(null);
+    }
+
+    private Map<String, Object> getCachedSchemaProperties(CatalogSchemaName catalogSchemaName)
+    {
+        IcebergMetadata icebergMetadata = metadataFactory.create(null);
+        return icebergMetadata.getSchemaProperties(null, catalogSchemaName);
+    }
+
+    private Optional<TrinoPrincipal> getCachedSchemaOwner(CatalogSchemaName catalogSchemaName)
+    {
+        IcebergMetadata icebergMetadata = metadataFactory.create(null);
+        return icebergMetadata.getSchemaOwner(null, catalogSchemaName);
+    }
+
+    private Table getCachedTableHandle(SchemaTableName schemaTableName)
+    {
+        try {
+            TrinoCatalog catalog = metadataFactory.getCatalogFactory().create(null);
+            if (catalogType.equals(CatalogType.GLUE)) {
+                return ((TrinoGlueCatalog) catalog).loadTableWithoutCache(null, schemaTableName);
+            }
+            else {
+                return catalog.loadTable(null, schemaTableName);
+            }
+        }
+        catch (TableNotFoundException e) {
+            return null;
```

Comment:
Returning null from a CacheLoader will result in an InvalidCacheLoadException. In case of a "CREATE TABLE" query, the loadTable function is throw to return TableNotFound exception, and thus will return null, so DDL queries will fail. We should not catch the exception here and let the "getTableHandle" function catch the exception and return null.

Reply:
As per (this doc)[https://guava.dev/releases/19.0/api/docs/com/google/common/cache/CacheLoader.html], on throwing an exception from inside a CacheLoader function the value is not cached. So, we are not caching a non-existent table, instead just throwing an exception and letting the "getTableHandle" function know that the table is not found. (Edited my above comment, maybe i was not clear enough)
So, in that case, are the non-existent tables being cached ? (maybe i am missing something here)

Reply:
Have removed additional exception handling from getCachedTableHandle.